---
title: "Introduction"
---

# Introduction

### Eh... ORMAN?

ORMAN is a tool for resolving multi-mappings within an RNA-Seq SAM file. ORMAN was publised in [Bioinformatics in October 2013][3].

### How do I get ORMAN?

Just clone our repository and issue `make` command:

	git clone https://github.com/sfu-compbio/orman.git
	cd orman && make -j

This will work only if you have your CPLEX installed in `/opt/ibm/cplex12_5_1`. If not, keep reading, or just try out our x86-64 [binary](https://github.com/sfu-compbio/orman/tarball/binary) (compiled on CentOS 6.x).

### How do I get CPLEX?

You will need to install IBM CPLEX Optimization Studio (preferably versions >= 12.5.1) before compiling ORMAN. You can get the trial version at [IBM's website](http://www-01.ibm.com/software/websphere/products/optimization/cplex-studio-preview-edition/).

If you are student or a faculty member, you can get CPLEX for free (woohoo!). Just sign up for [Academic Initiative Program](http://www-03.ibm.com/ibm/university/academic/pub/page/academic_initiative).

After you are done with installing your CPLEX, please set the `CPLEXDIR` variable to point to the CPLEX installation directory. For example, if you have installed CPLEX in `~/apps/mycplex`, run `make` as follows:

	CPLEXDIR=~/apps/mycplex make

If you are not using x86-64 system, you will also need to set up `CPLEXARCH` variable. Please refer to the CPLEX documentation to discover its proper value (but please note that we cannot guarantee that ORMAN will run properly on such systems).

### How do I run ORMAN?

ORMAN is invoked as following:

	orman -t [thread count] -g [ORMAN gtf file] -s [sam file] [output file]

Example invocation (results will be stored in `orman_awesome.sam` file):

	orman -t 10 -g super.gtf -s awesome.sam orman_awesome.sam

**Please note that input SAM file should be sorted by the read name, and it should have valid headers.**

#### ormanGTF

ORMAN supports only GTF files containing ORMAN-specific additional information (namely `partial_ex` field). If you don't have such file, use the provided `ormanGTF` script to create compatinge GTF file as follows:
	
	ormanGTF < [input gtf file] > [new gtf file]

Example:

	ormanGTF < almost_super.gtf > super.gtf

[3]: http://bioinformatics.oxfordjournals.org/content/30/5/644
